### PR TITLE
refactor: Refactors InSlot classes

### PR DESCRIPTION
**Pull request by vdiezel**
according to what we discussed at the board.

@DanielSeehausen  start review here: https://github.com/ospin-web-dev/FCTGraph/pull/35/commits/20c9344d4baeef6f88aa30f8d65bf9e8580fa878

### CODE REVIEW

**Review by DanielSeehausen — CHANGES_REQUESTED** [review]
have a few blocking requests have to run now though will add them tomorrow! including:
seeders, removing certain stuff from the virtual class now, documentation (readme)

File: src/seeders/functionalities/UnknownActuatorSeeder.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
-const SensorSeeder = require('./SensorSeeder')
+const ActuatorSeeder = require('./ActuatorSeeder')
 const { InSlotSeeder } = require('../slots')
 const UnknownActuator = require('../../functionalities/UnknownActuator')
 
-class UnknownActuatorSeeder extends SensorSeeder {
+class UnknownActuatorSeeder extends ActuatorSeeder {
```
Comment (DanielSeehausen) [inline src/seeders/functionalities/UnknownActuatorSeeder.js:5]: 👀

File: src/seeders/slots/InSlotSeeder.js [hunk 2]
```diff
@@ -78,29 +75,26 @@ class InSlotSeeder extends SlotSeeder {
...
+        BooleanInSlot.DATA_TYPE,
+        OneOfInSlot.DATA_TYPE,
+      ])
 
     /* this method will delegate and attempt to build fake data based
      * off of what the user has provided. e.g. if the user provided a
      * max value, it will try to give sensible min and defaultValues.
      * the reason all of these values are generated at once is because
      * they are all, per the JOI SCHEMA, related
      */
```
Comment (DanielSeehausen) [inline src/seeders/slots/InSlotSeeder.js:91]: this comment seems to have gotten estranged from the method it was providing info for

File: src/seeders/slots/InSlotSeeder.js [hunk 3]
```diff
@@ -78,29 +75,26 @@ class InSlotSeeder extends SlotSeeder {
...
   static generate(data = {}) {
     const slotData = super.generate(data)
-    slotData.dataType = data.dataType || faker.random.arrayElement(Object.values(InSlot.DATA_TYPES))
+    slotData.dataType = data.dataType || faker.random
+      .arrayElement([
+        FloatInSlot.DATA_TYPE,
+        IntegerInSlot.DATA_TYPE,
+        BooleanInSlot.DATA_TYPE,
+        OneOfInSlot.DATA_TYPE,
+      ])
```
[the remark this answers is not in the document]
Comment (DanielSeehausen) [reply src/seeders/slots/InSlotSeeder.js:84]: The new inslot classes need seeders, which extend the now virtual `InSlotSeeder`

it was for the annoying min max etc. behavior that is similar in float and integer sub classes that I suggested that additional intermediate layer of a virtual class (which would come with a virtual seeder). I think there is a clean way around it though - in the `Integer/FloatInSlotSeeder`s one should be able to first define the `defaultValue` and anything else that this dynamic seed random min/max value needs to run. Then when `super` is called, that can be passed, so the virtual `InSlotSeeder` can use that to derive the min/max if they aren't provided!

**Review by DanielSeehausen — CHANGES_REQUESTED** [review]
I have to run to the train I havent finished reviewing yet - will asap on train.

I do have a blocking request re: the seeder. See how the functionalities are structure (e.g. `FunctionalitySeeder` -> `SensorSeeder` -> `TemperatureSensorSeeder`) to ensure `InSlotSeeder` is virtual

File: src/slots/BooleanInSlot.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+const JOIous = require('../mixins/instanceMixins/JOIous')
+const InSlot = require('./InSlot')
+
+class BooleanInSlot extends InSlot {
+
+  constructor({ defaultValue = null, ...slotData }) {
+    super(slotData)
+    this.dataType = BooleanInSlot.DATA_TYPE
+    this.defaultValue = defaultValue
+  }
```
Comment (DanielSeehausen) [inline src/slots/BooleanInSlot.js:12]: the `constructor` is an instance method,
which is why, I assume, you tend to see it defined below the `static` methods with the other instance methods. Certainly not a blocking request. I did not open this comment to write that though, so I digress.

Should `BooleanInSlot`s (and others) ever only be `tareable: true`? If so, do you think it would be appropriate to update the `SCHEMA` to reflect that?

File: src/seeders/slots/InSlotSeeder.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (DanielSeehausen) [reply src/seeders/slots/InSlotSeeder.js:84]: Now that I have read more of the PR, it seems you already have one foot in that door, seeing how the subclasses (e.g. `FloatInSlot`) now own their min/max reality instead of the `InSlot`

File: src/slots/InSlot.js [hunk 5]
```diff
@@ -9,131 +8,32 @@ class InSlot extends Slot {
...
-    this.min = min
-    this.max = max
-    this.selectOptions = selectOptions
-  }
-
-  serializeOptionals() {
-    // class method useful for seeding
-    // which is why the instance is delegating up
-    return InSlot.SERIALIZE_OPTIONAL_DATA(this)
   }
```
Comment (DanielSeehausen) [inline src/slots/InSlot.js:25]: nice that this got rid of this jank!

File: src/slots/InSlot.js [hunk 6]
```diff
@@ -9,131 +8,32 @@ class InSlot extends Slot {
...
-      if (value !== undefined) obj[key] = value
-      return obj
-    }, {})
-  }
-
   static get SCHEMA() {
     return Joi.object({
       type: Joi.string().allow(InSlot.TYPE).required(),
-      dataType: Joi.string().allow(...Object.values(InSlot.DATA_TYPES)).required(),
       tareable: Joi.boolean().required(),
```
Comment (DanielSeehausen) [inline src/slots/InSlot.js:14]: can all `InSlot`s be `tareable`?

File: src/slots/factories/SlotFactory.js [hunk 7]
```diff
@@ -15,15 +27,48 @@ class SlotFactory {
...
 
-  static get TYPE_TO_CLASS() { return ({ InSlot, OutSlot }) }
+  static get DATATYPE_TO_IN_SLOT_CLASS() {
+    return {
+      [IntegerInSlot.DATA_TYPE]: IntegerInSlot,
+      [FloatInSlot.DATA_TYPE]: FloatInSlot,
+      [BooleanInSlot.DATA_TYPE]: BooleanInSlot,
+      [OneOfInSlot.DATA_TYPE]: OneOfInSlot,
+    }
+  }
```
Comment (DanielSeehausen) [inline src/slots/factories/SlotFactory.js:37]: NOICE

File: src/slots/factories/SlotFactory.js [hunk 8]
```diff
@@ -15,15 +27,48 @@ class SlotFactory {
...
+
+  static get DATATYPE_TO_OUT_SLOT_CLASS() {
+    return {
+      [OutSlot.DATA_TYPES.INTEGER]: OutSlot,
+      [OutSlot.DATA_TYPES.FLOAT]: OutSlot,
+    }
+  }
```
Comment (DanielSeehausen) [inline src/slots/factories/SlotFactory.js:44]: we very well may want/need/decide to do the same thing with OutSlots that you are doing now with InSlots in the future and this sets that potential change up really nicely! kind programming.

File: src/slots/factories/SlotFactory.js [hunk 9]
```diff
@@ -15,15 +27,48 @@ class SlotFactory {
...
+    const { type, dataType } = slotData
+    const FoundClass = SlotFactory.getSlotClass(slotData)
 
     if (!SlotFactory.classIsSupported(FoundClass)) {
       const json = JSON.stringify(slotData)
-      throw new Error(`Slot type not supported ${type || 'FALSEY'}\n${json}`)
+      throw new Error(`No ${dataType} slot class found for ${type}\n${json}`)
     }
 
     return new FoundClass(slotData)
```
Comment (DanielSeehausen) [inline src/slots/factories/SlotFactory.js:74]: you extended this file like an artist. super nice

File: test/slots/BooleanInSlot.test.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
+const BooleanInSlot = require('slots/BooleanInSlot')
+const { InSlotSeeder } = require('seeders/slots')
+/* the SlotSeeder is not present in the seeders/slots index because it
+ * should not be exposed as public seeder interface */
```
Comment (DanielSeehausen) [inline test/slots/BooleanInSlot.test.js:4]: this comment can go

File: test/slots/FloatInSlot.test.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+    it('throws when defaultValue is below min', () => {
+      const slotData = InSlotSeeder.generate({ dataType: FloatInSlot.DATA_TYPE })
+      slotData.min = 300.1
+      slotData.defaultValue = 200.1
+
+      expect(() => new FloatInSlot(slotData)).toThrow(/min/)
+    })
+  })
+})
```
Comment (DanielSeehausen) [inline test/slots/FloatInSlot.test.js:87]: noice

File: test/slots/factories/SlotFactory.test.js [hunk 12]
```diff
@@ -21,8 +25,22 @@ describe('the slot factory', () => {
...
+    })
+
     describe('when making InSlots', () => {
-      Object.values(InSlot.DATA_TYPES).forEach(dataType => {
+      [
+        FloatInSlot.DATA_TYPE,
+        IntegerInSlot.DATA_TYPE,
+        BooleanInSlot.DATA_TYPE,
+        OneOfInSlot.DATA_TYPE,
+      ].forEach(dataType => {
```
Comment (DanielSeehausen) [inline test/slots/factories/SlotFactory.test.js:43]:
```suggestion
      const supportedDataTypes = SlotFactory.SUPPORTED_CLASSES.map(
        SlotClass => SlotClass.DATA_TYPE
      )
      
      supportedDataTypes.forEach(dataType => {
```

may make the test less brittle and/or auto extend it should someone add a new dataType

File: test/slots/factories/SlotFactory.test.js [hunk 13]
```diff
@@ -21,8 +25,22 @@ describe('the slot factory', () => {
       }).toThrow('Slot type not supported FALSEY')
     })
 
+    it('throws error when the type does not have a class defined for the provided dataType', () => {
+      const inSlotData = InSlotSeeder.generate()
+      const bogusDataType = 'trump'
```
[the remark this answers is not in the document]
Comment (DanielSeehausen) [reply test/slots/factories/SlotFactory.test.js:30]: image

**Comment by DanielSeehausen** [comment]
I finished reading - PR looks GREAT. We are definitely seeing the same beans when looking at this code.

I only have one blocking request, and that's to update the slot seeders. (If you wanted, you could also then expose them in the public interface and update the readme, but that is perhaps a YAGNI at this time?)
As far as non-blocking changes go, I left a small question re: 'tareable' and the various new `InSlot` types

---

In summary, I am 95% sure that you only had to do all this work because I was irresponsible and did not do what the little voice was telling me to at the time of writing this (re: subclassing out `InSlot`). I am kicking myself a little for not just doing it at the start.

File: src/slots/InSlot.js (hunk 6 above)
Comment (vdiezel) [reply src/slots/InSlot.js:14]: I was back and forth on this.  Probably only the integer and float. I'll change that (will require a couple of PRs on hambda to fix that)

File: src/seeders/slots/InSlotSeeder.js (hunk 2 above)
Comment (vdiezel) [reply src/seeders/slots/InSlotSeeder.js:91]: removed it

File: src/slots/InSlot.js (hunk 6 above)
Comment (vdiezel) [reply src/slots/InSlot.js:14]: done

File: src/slots/BooleanInSlot.js (hunk 4 above)
Comment (vdiezel) [reply src/slots/BooleanInSlot.js:12]: changed the order; removed tareable from boolean and oneOf slots.

File: test/slots/BooleanInSlot.test.js (hunk 10 above)
Comment (vdiezel) [reply test/slots/BooleanInSlot.test.js]: removed

File: src/seeders/slots/BooleanInSlotSeeder.js [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    const slotData = super.generate(data)
+    slotData.dataType = BooleanInSlot.DATA_TYPE
+
+    return {
+      ...slotData,
+      defaultValue: BooleanInSlotSeeder.generateDefaultValue(slotData),
+    }
+  }
+
+}
```
Comment (DanielSeehausen) [inline src/seeders/slots/BooleanInSlotSeeder.js:25]:
```suggestion
class BooleanInSlotSeeder extends InSlotSeeder {

  static generate(data = {}) {
    const slotData = super.generate(data)

    return {
      ...slotData,
      dataType: BooleanInSlot.DATA_TYPE,
      defaultValue: faker.datatype.boolean(),
      ...data,
    }
  }

}
```

File: src/seeders/slots/FloatInSlotSeeder.js [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+  static generate(data = {}) {
+    const slotData = super.generate(data)
+    slotData.dataType = FloatInSlot.DATA_TYPE
+
+    return {
+      ...slotData,
+      tareable: faker.datatype.boolean(),
+      ...FloatInSlotSeeder.generateMinMaxDefaultValue(slotData),
+    }
+  }
```
Comment (DanielSeehausen) [inline src/seeders/slots/FloatInSlotSeeder.js:46]: same request re: allowing data override and not assigning to the parent data

File: src/seeders/slots/IntegerInSlotSeeder.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  static generate(data = {}) {
+    const slotData = super.generate(data)
+    slotData.dataType = IntegerInSlot.DATA_TYPE
+
+    return {
+      ...slotData,
+      tareable: faker.datatype.boolean(),
+      ...IntegerInSlotSeeder.generateMinMaxDefaultValue(slotData),
+    }
+  }
```
Comment (DanielSeehausen) [inline src/seeders/slots/IntegerInSlotSeeder.js:46]: same request re: allowing data override and not assigning to the parent data

File: src/seeders/slots/OneOfInSlotSeeder.js [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+  static generate(data = {}) {
+    const slotData = super.generate(data)
+    slotData.dataType = OneOfInSlot.DATA_TYPE
+
+    return {
+      ...slotData,
+      ...OneOfInSlotSeeder.generateSelectOptionsAndDefaultValue(slotData),
+    }
+  }
+
```
Comment (DanielSeehausen) [inline src/seeders/slots/OneOfInSlotSeeder.js:30]: same request re: allowing data override and not assigning to the parent data

File: test/slots/BooleanInSlot.test.js [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    const forbiddenKeys = ['min', 'max', 'selectOptions']
+
+    it('creates a boolean slot with the expected properties', () => {
+      const slotData = BooleanInSlotSeeder.generate()
+
+      const slot = new BooleanInSlot(slotData)
+
+      requiredKeys.forEach(key => expect(key in slot).toBe(true))
+      forbiddenKeys.forEach(key => expect(key in slot).toBe(false))
+    })
```
Comment (DanielSeehausen) [inline test/slots/BooleanInSlot.test.js:16]: nice simple extensible setup

File: test/slots/Slot.test.js [hunk 19]
```diff
@@ -28,33 +34,28 @@ describe('the Slot class', () => {
...
+        const slotData = IntegerInSlotSeeder.generate()
+        delete slotData.tareable
 
-        const slot = new InSlot(slotData)
+        const slot = new IntegerInSlot(slotData)
 
-        expect(slot.defaultValue).toBeNull()
+        expect(slot.tareable).toBe(false)
       })
     })
```
Comment (DanielSeehausen) [inline test/slots/Slot.test.js:57]: im not 100% sure about the following comment:

I believe `InSlot` no longer has anything to do with `tareable`
Both of these seem to actually be testing behavior the `Float/IntegerInSlot.constructor` are doing, and not the `Slot.constructor`.

File: test/slots/Slot.test.js [hunk 20]
```diff
@@ -86,7 +87,7 @@ describe('the Slot class', () => {
   describe('.assertStructure', () => {
 
     it('blows up because Slot is a virtual class and it wants to kindly tell you that a mistake was likely made in a child that has not defined the method', () => {
-      const slot = new Slot(InSlotSeeder.generate())
+      const slot = new Slot(FloatInSlotSeeder.generate())
```
Comment (DanielSeehausen) [inline test/slots/Slot.test.js:90]: This may be sufficient and closer to what is being tested. `seedOne` is one of the extended methods from `FactorySeeder` which all the seeders have. It runs the seeder's `generate` followed by the seeders `SEED_METHOD` (which, in this case, is `new Slot`)

```suggestion
      const slot = SlotSeeder.seedOne()
```

File: test/slots/Slot.test.js (hunk 19 above)
Comment (vdiezel) [reply test/slots/Slot.test.js:57]: this is correct, I forgot to move those to the particular single slot test file

Comment (vdiezel) [reply test/slots/Slot.test.js:57]: fixed

**Review by DanielSeehausen — APPROVED** [review]
looks good - just what was discussed and additional small comments below!

**Review by DanielSeehausen — APPROVED** [review]
